Fix BigQuery Percentile and Median expressions not using correct backtick quoting

jeff303 opened the pull request:
Ensure the identifier components are quoted within the `:percentile` hsql (raw) form

Add test to ensure query is created properly

Mark `:percentile-aggregations` as being supported in the new driver

jeff303 committed:
Fix BigQuery Percentile and Median expressions not using correct backtick quoting

Ensure the identifier components are quoted within the `:percentile` hsql (raw) form

Add test to ensure query is created properly

Mark :percentile-aggregations as being supported in the new driver

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -419,7 +419,9 @@
   [driver [_ arg p]]
   (let [[offset quantiles] (percentile->quantile p)]
     (hsql/raw (format "APPROX_QUANTILES(%s, %s)[OFFSET(%s)]"
-                      (hformat/to-sql (sql.qp/->honeysql driver arg))
+                      (-> (sql.qp/->honeysql driver arg)
+                          (hformat/format :quoting (sql.qp/quote-style driver))
+                          first)
```
The right way to do this would be to have this function return a reified `ToSql`, here's an example: https://github.com/metabase/metabase/blob/d2433696380fb43bb2ffaf5e1b2aabebd1354bc2/src/metabase/driver/postgres.clj#L186-L197

We don't need to use `hsql/raw` here. In the impl of `to-sql` then we can recursively apply `to-sql` to its args

jeff303 replied:
Made the change.

camsaul reviewed (changes requested):
The right way to fix this is to `reify` `ToSql`... calling HoneySQL `format` isn't something that should be done as part of _HoneySQL_ generation itself.

Query compilation is basically

1.  MBQL transformations in QP middleware
2. Compile MBQL to HoneySQL map in SQL QP & driver methods
3. Compile HoneySQL map to SQL string -- mostly done in HoneySQL itself but done in a custom manner for a few reified `ToSQL` types and ` hformat/fn-handler` method impls

Using `hsql/raw` and HoneySQL `format` in step 2 is doing things prematurely and muddying up the distinction between the two steps... the HoneySQL->SQL compilation should happen all at once in step 3

jeff303 committed:
PR feedback

reify ToSql instead of mixing up HoneySQL forms with formatting

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -415,13 +416,25 @@
+    hformat/ToSql
+    (to-sql [_]
+      (format "APPROX_QUANTILES(%s, %s)[OFFSET(%s)]"
+        (hformat/to-sql (sql.qp/->honeysql :bigquery-cloud-sdk expr))
```
`->honeysql` should probably happen inside `approx-quantities` itself. `approx-quantities` is part of the "generate HoneySQL" stage, so we should do all the HoneySQL generation there. Then we can just call `to-sql` on the generated HoneySQL.

Also, this is a small nitpick, but it would be better to make `driver` an explicit parameter rather than hardcoding `:bigquery-cloud-sdk`. The reason is that if we ever create a new driver derived from `:bigquery-cloud-sdk` then it would potentially not be using the correct method implementation for `->honeysql`.

jeff303 replied:
Fixed in the latest commit.

camsaul reviewed (approved):
A few minor notes but other than that LGTM

jeff303 committed:
More PR feedback

Capture HoneySQL form for expr outside of ToSql reification

Add driver as parameter
